plugins/preview/plugin.js

```diff
@@ -8,128 +8,262 @@
+	'use strict';
 
-	var previewCmd = { modes: { wysiwyg: 1, source: 1 },
-		canUndo: false,
-		readOnly: 1,
-		exec: function( editor ) {
-			var sHTML,
-				config = editor.config,
-				baseTag = config.baseHref ? '<base href="' + config.baseHref + '"/>' : '',
-				eventData;
-
-			if ( config.fullPage )
-				sHTML = editor.getData().replace( /<head>/, '$&' + baseTag ).replace( /[^>]*(?=<\/title>)/, '$& &mdash; ' + editor.lang.preview.preview );
-			else {
-				var bodyHtml = '<body ',
-					body = editor.document && editor.document.getBody();
-
-				if ( body ) {
-					if ( body.getAttribute( 'id' ) )
-						bodyHtml += 'id="' + body.getAttribute( 'id' ) + '" ';
-					if ( body.getAttribute( 'class' ) )
-						bodyHtml += 'class="' + body.getAttribute( 'class' ) + '" ';
-				}
-
-				bodyHtml += '>';
-
-				sHTML = editor.config.docType + '<html dir="' + editor.config.contentsLangDirection + '">' +
-					'<head>' +
-						baseTag +
-						'<title>' + editor.lang.preview.preview + '</title>' +
-						CKEDITOR.tools.buildStyleHtml( editor.config.contentsCss ) +
-					'</head>' + bodyHtml +
-						editor.getData() +
-					'</body></html>';
+	var pluginName = 'preview',
+		previewCmd;
+
+	// Register a plugin named "preview".
+	CKEDITOR.plugins.add( pluginName, {
+		// jscs:disable maximumLineLength
+		lang: 'af,ar,az,bg,bn,bs,ca,cs,cy,da,de,de-ch,el,en,en-au,en-ca,en-gb,eo,es,es-mx,et,eu,fa,fi,fo,fr,fr-ca,gl,gu,he,hi,hr,hu,id,is,it,ja,ka,km,ko,ku,lt,lv,mk,mn,ms,nb,nl,no,oc,pl,pt,pt-br,ro,ru,si,sk,sl,sq,sr,sr-latn,sv,th,tr,tt,ug,uk,vi,zh,zh-cn', // %REMOVE_LINE_CORE%
+		// jscs:enable maximumLineLength
+		icons: 'preview,preview-rtl', // %REMOVE_LINE_CORE%
+		hidpi: true, // %REMOVE_LINE_CORE%
+		init: function( editor ) {
+			var pluginPath = CKEDITOR.plugins.getPath( 'preview' );
+
+			editor.addCommand( pluginName, previewCmd );
+			editor.ui.addButton && editor.ui.addButton( 'Preview', {
+				label: editor.lang.preview.preview,
+				command: pluginName,
+				toolbar: 'document,40'
+			} );
+
+			if ( !CKEDITOR.env.webkit || CKEDITOR.env.chrome ) {
+				return;
 			}
 
-			var iWidth = 640,
-				// 800 * 0.8,
-				iHeight = 420,
-				// 600 * 0.7,
-				iLeft = 80; // (800 - 0.8 * 800) /2 = 800 * 0.1.
-			try {
-				var screen = window.screen;
-				iWidth = Math.round( screen.width * 0.8 );
-				iHeight = Math.round( screen.height * 0.7 );
-				iLeft = Math.round( screen.width * 0.1 );
-			} catch ( e ) {}
+			// If Safari does not have print styles in cache, it won't
+			// render the printed content. Therefore we need to preload the styles.
+			preloadStyles( editor.config.contentsCss );
+			preloadStyles( pluginPath + 'styles/screen.css' );
+		}
+	} );
+
+	/**
+	 * Namespace providing a set of helper functions for working with print preview, exposed by the
+	 * [Preview](https://ckeditor.com/cke4/addon/tableselection) plugin.
+	 *
+	 * @since 4.14.0
+	 * @singleton
+	 * @class CKEDITOR.plugins.preview
+	 */
+	CKEDITOR.plugins.preview = {
```

Comment:
Refactoring nicely simplified this module. I'm only wondering if we could go a bit further, and use here our good old `CKEDITOR.tools.createClass` for even better readability (and start using OOP finally in our plugins). I already did something like that during working on quick table plugin and it seems to improve a bit implementation:
https://github.com/ckeditor/ckeditor4/blob/c23ca76d0cf074e713094a074274f2211cb931bb/plugins/table/plugin.js#L166
https://github.com/ckeditor/ckeditor4/blob/c23ca76d0cf074e713094a074274f2211cb931bb/plugins/quicktable/plugin.js#L26

It would be nice to start using a single adopted method for modules. WDYT?

Reply:
Nearly every plugin uses namespaces to just store some helper functions, often unrelated with each other, e.g. https://github.com/ckeditor/ckeditor4/blob/37e9cb17403336a06f1065d83be376142fd5cb9d/plugins/tableselection/plugin.js#L941-L958. Making it a class, which members would basically share only `editor` property, doesn't sound like OOP to me.

OTOH these namespaces exist only to allow better testing or ability to override some parts of plugins from other plugins (as it was done in Table Selection, which reuses and overrides some parts of Table Tools). They do not serve any other purpose.

Reply:
> Making it a class, which members would basically share only editor property, doesn't sound like OOP to me.

OOP is not only about sharing some common properties or state, but also encapsulate domain logic in unit boundaries. Although:

>OTOH these namespaces exist only to allow better testing or ability to override some parts of plugins from other plugins (as it was done in Table Selection, which reuses and overrides some parts of Table Tools). They do not serve any other purpose.

Let's keep it simple then:+1: and worry if it ever grows.